feat: Deployments page

deboer-tim opened the pull request:
### What does this PR do?

Initial Deployments page. Creates a new route and icon in the main nav, plus
a regular page/table with deployment details. All columns from the design are
included, along with delete action and multiple deletion.

Notes:
- Deployments are not refreshing. It is likely unrelated to this PR, but
  should be resolved before this PR is merged. If you have existing
  deployments or Force Refresh everything else is working. (Fixed via #4989)
- In the design the Pods column are dots like the Pod list containers. This
  is deferred to a separate PR.
- We could sort by more columns... but not critical/can be done later.
- We should hide the nav item when there are no active Kubernetes providers. (PR updated to hide Deployments until you have a Kubernetes context)

### Screenshot/screencast of this PR

<img width="1060" alt="Screenshot 2023-11-23 at 3 16 18 PM" src="https://github.com/containers/podman-desktop/assets/19958075/0fd04aa5-d090-47d6-bb5a-0ec4561514e6">

### What issues does this PR fix or reference?

Final major part of #4366.

### How to test this PR?

Set kube context to a cluster with a deployment.

deboer-tim commented:
Pushed new branch based on main with the final changes, ready for regular review.

lstocchi commented:
Played with it and it looks nice. I'll give my LGTM after we have https://github.com/containers/podman-desktop/issues/4974 otherwise it looks like it doesn't work. If you delete a deploy the table does not get refreshed, ... I'll give it a look for the ingress part

lstocchi commented:
Other thing is, should we display some message if there are no deployments? This is what i see now
![image](https://github.com/containers/podman-desktop/assets/49404737/ceb24327-264c-4d00-950e-57252e2ae8c6)

deboer-tim commented:
Added new commit. Here's my view on where things stand:
- We should not merge until there is a fix for #4974. I've put up PR #4989 that solves the problem for deployments, but we need to merge that or a broader solution.
- The new commit hides the Deployment icon from the main nav when there are no Kubernetes contexts. I think this is a reasonable simple test for now, since we want to keep the UI simple when you aren't using Kubernetes, and when you are you could have a provider within Podman Desktop but could just have created a context externally. 
- The commit adds the missing empty screens for no deployments or all deployments filtered by search. There is no need
for a 'no provider' screen like other list pages because of the nav hiding.
- The commit also adds tests that were missing - basic UI and empty screens.

jeffmaury commented on packages/renderer/src/lib/deployments/DeploymentColumnNamespace.svelte:
```diff
@@ -0,0 +1,9 @@
+<script lang="ts">
```
I'm wondering why we don't have a generic NamespaceColumn as it's likely that we will have a lot of Kubernetes objects that are namespace aware

benoitf commented on packages/renderer/src/lib/deployments/DeploymentUI.ts:
```diff
@@ -18,9 +18,17 @@
+
+export interface DeploymentCondition {
+  type: string;
+  message: string | undefined;
```
I think you need to review other places where we need to use the optional `?` vs `| undefined`
```
message? : string
```

benoitf commented on packages/renderer/src/lib/deployments/DeploymentUI.ts:
```diff
@@ -18,9 +18,17 @@
   namespace: string;
   replicas: number;
   ready: number;
+  created: Date | undefined;
```
```suggestion
  created?: Date;
```

deboer-tim replied:
Fixed in new commit.

benoitf commented on packages/renderer/src/lib/deployments/DeploymentUI.ts:
```diff
@@ -18,9 +18,17 @@
+
+export interface DeploymentCondition {
+  type: string;
+  message: string | undefined;
```
```suggestion
  message?: string;
```

deboer-tim replied:
Fixed in new commit.

benoitf commented on packages/renderer/src/lib/deployments/deployment-utils.ts:
```diff
@@ -29,16 +29,47 @@ export class DeploymentUtils {
+
   getDeploymentUI(deployment: V1Deployment): DeploymentUI {
+    const conditions = [];
+    if (deployment.status?.conditions) {
+      for (const con of deployment.status.conditions) {
+        const c: DeploymentCondition = {
+          type: con.type,
+          message: con.message,
+        };
+        conditions.push(c);
+      }
+    }
```
should be simplified using map like approach or reduce like approach
```suggestion
conditions = (deployment.status.conditions ?? []).map(c => {
          type: con.type,
          message: con.message,
        }
);
```

benoitf commented on packages/renderer/src/lib/deployments/deployment-utils.ts:
```diff
@@ -29,16 +29,47 @@ export class DeploymentUtils {
       namespace: deployment.metadata?.namespace || '',
-      age: this.humanizeAge((deployment.metadata?.creationTimestamp || '').toString()),
+      created: deployment.metadata?.creationTimestamp,
+      age: deployment.metadata?.creationTimestamp
+        ? this.humanizeAge(deployment.metadata.creationTimestamp.toString())
```
can use maybe the same approach then previous lines with ?? or || if there is no value
or we let humanizeAge accept undefined and it returns '' in that case

deboer-tim replied:
I tried a couple ways and went with humanizeAge() accepting undefined since it seemed like the shortest/simplest code to me.

deboer-tim commented:
I think I've addressed all the issues raised, but this PR is stagnating. Could I please get another review or comment on what else this needs?

cdrage reviewed (changes requested):
After my review, I realized while I was playing more with it that there is no sorting for Pods or Age (I understand why we cant sort Conditions)?

<img width="1342" alt="Screenshot 2023-11-30 at 10 32 49 PM" src="https://github.com/containers/podman-desktop/assets/6422176/eec01f37-696a-4c88-830f-745e16baf08c">

deboer-tim commented:
Sorry, I should have explained - I was worried about both the size of this PR and it dragging on, so I was trying to find a reasonable spot to stop and why it says "Final major part of 4366" and not "Fixes".

You've listed exactly what I still planned to do: we should have pod linking and column sorting, and the design has the pods column like the pod view's container dots. If you think this isn't a good place to merge I can withdraw and include more of that here.

benoitf commented on packages/renderer/src/lib/deployments/DeploymentColumnAge.svelte:
```diff
@@ -0,0 +1,9 @@
+<script lang="ts">
+import type { DeploymentUI } from './DeploymentUI';
+
+export let object: DeploymentUI;
+</script>
+
+<div class="text-xs text-gray-500">
+  {object.age}
+</div>
```
now it seems you can use your custom rendering / simple column

deboer-tim replied:
Rebased on main to pick this up, and replaced Age and Namespace columns.

benoitf commented on packages/renderer/src/lib/deployments/DeploymentColumnStatus.svelte:
```diff
@@ -0,0 +1,9 @@
+<script lang="ts">
+import StatusIcon from '../images/StatusIcon.svelte';
+import DeploymentIcon from '../images/DeploymentIcon.svelte';
+import type { DeploymentUI } from './DeploymentUI';
+
+export let object: DeploymentUI;
+</script>
+
+<StatusIcon icon="{DeploymentIcon}" status="{object.status}" />
```
missing a component test

deboer-tim replied:
Added tests for all columns.

benoitf commented on packages/renderer/src/lib/deployments/DeploymentEmptyScreen.svelte:
```diff
@@ -0,0 +1,9 @@
+import DeploymentIcon from '../images/DeploymentIcon.svelte';
+</script>
+
+<EmptyScreen
```
missing a component test

deboer-tim replied:
Added tests for all columns.

benoitf commented on packages/renderer/src/lib/deployments/deployment-utils.ts:
```diff
@@ -22,23 +22,48 @@ import type { DeploymentUI } from './DeploymentUI';
+      return '';
+    }
+    // make it human friendly
+    return this.humanizeAge(deployment.created.toString());
```
I'm wondering why we use the toString and do not send the `.created` field directly

humanizeAge should be smarter

deboer-tim replied:
Yes, this was poor. Updated humanizeAge to accept Date.

benoitf commented on packages/renderer/src/lib/deployments/deployment-utils.ts:
```diff
@@ -22,23 +22,48 @@ import type { DeploymentUI } from './DeploymentUI';
       namespace: deployment.metadata?.namespace || '',
-      age: this.humanizeAge((deployment.metadata?.creationTimestamp || '').toString()),
+      created: deployment.metadata?.creationTimestamp,
+      age: this.humanizeAge(deployment.metadata?.creationTimestamp?.toString()),
```
here I would not use `.toString`, send `.creationTImestamp`

deboer-tim replied:
Fixed.

deboer-tim commented:
Rebased on main and responded to feedback. Added default name sorting and replaced Age/Namespace columns now that there are new options. Added tests, fixed humanizeAge(), and made the Age column sortable.

As per my previous comment, I felt like this was a reasonable place to merge. In a future PR I'd make the Pods column into dots and allow sorting there. Please re-review or let me know if you disagree. Thanks.

cdrage reviewed (approved):
LGTM! Amazing.

Obviously we should merge after 1.6 / finalization of Icon?

deboer-tim commented:
The deployment icon is finalized (unless we get feedback in future), it's just the Ingress/Route icon which Conor was working on (and he pinged me a final one overnight).

Yes, given the size I would expect to wait for another approval + 1.6.

deboer-tim commented:
@jeffmaury or @benoitf ?

benoitf reviewed (approved):
I thought it was waiting the Informer stuff

deboer-tim commented:
A deployment store with a regular Kube watcher (same as we do for pods) was merged before Luca came up with the informer idea. So it works fine, but I've started a separate PR to refactor deployment store to informer.
